File: app/src/main/java/com/example/prj_gifu_univ_bus_navi/data/LocalAcademicCalendarData.java
package com.example.prj_gifu_univ_bus_navi.data;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public final class LocalAcademicCalendarData {
    private static final List<DateRange> CLASS_PERIODS = Collections.unmodifiableList(Arrays.asList(
        new DateRange(LocalDate.of(2026, 4, 10), LocalDate.of(2026, 8, 6)),
        new DateRange(LocalDate.of(2026, 10, 1), LocalDate.of(2027, 2, 5))
    ));
    private static final List<LocalDate> INDIVIDUAL_HOLIDAYS = Collections.unmodifiableList(buildIndividualHolidays());

    private LocalAcademicCalendarData() {
    }

    public static boolean isClassDay(LocalDate date) {
        return isInClassPeriod(date)
            && date.getDayOfWeek() != DayOfWeek.SATURDAY
            && date.getDayOfWeek() != DayOfWeek.SUNDAY
            && !LocalHolidayData.isHoliday(date)
            && !INDIVIDUAL_HOLIDAYS.contains(date);
    }

    public static boolean isSchoolHoliday(LocalDate date) {
        return !isClassDay(date);
    }

    private static boolean isInClassPeriod(LocalDate date) {
        for (DateRange period : CLASS_PERIODS) {
            if (period.contains(date)) {
                return true;
            }
        }
        return false;
    }

    private static List<LocalDate> buildIndividualHolidays() {
        List<LocalDate> holidays = new ArrayList<>();
        holidays.add(LocalDate.of(2026, 10, 30));
        addRange(holidays, LocalDate.of(2026, 12, 26), LocalDate.of(2027, 1, 4));
        holidays.add(LocalDate.of(2027, 1, 15));
        return holidays;
    }

    private static void addRange(List<LocalDate> dates, LocalDate start, LocalDate endInclusive) {
        LocalDate current = start;
        while (!current.isAfter(endInclusive)) {
            dates.add(current);
            current = current.plusDays(1);
        }
    }

    private static final class DateRange {
        private final LocalDate start;
        private final LocalDate endInclusive;

        private DateRange(LocalDate start, LocalDate endInclusive) {
            this.start = start;
            this.endInclusive = endInclusive;
        }

        private boolean contains(LocalDate date) {
            return !date.isBefore(start) && !date.isAfter(endInclusive);
        }
    }
}

File: app/src/main/java/com/example/prj_gifu_univ_bus_navi/data/LocalAcademicCalendarData.kt
package com.example.prj_gifu_univ_bus_navi.data

import java.time.DayOfWeek
import java.time.LocalDate

object LocalAcademicCalendarData {
    private val classPeriods = listOf(
        DateRange(LocalDate.of(2026, 4, 10), LocalDate.of(2026, 8, 6)),
        DateRange(LocalDate.of(2026, 10, 1), LocalDate.of(2027, 2, 5)),
    )

    private val individualHolidays = buildSet {
        add(LocalDate.of(2026, 10, 30))
        addRange(LocalDate.of(2026, 12, 26), LocalDate.of(2027, 1, 4))
        add(LocalDate.of(2027, 1, 15))
    }

    fun isClassDay(date: LocalDate): Boolean =
        classPeriods.any { it.contains(date) } &&
            date.dayOfWeek !in setOf(DayOfWeek.SATURDAY, DayOfWeek.SUNDAY) &&
            date !in LocalHolidayData.holidays &&
            date !in individualHolidays

    fun isSchoolHoliday(date: LocalDate): Boolean = !isClassDay(date)

    private data class DateRange(
        val start: LocalDate,
        val endInclusive: LocalDate,
    ) {
        fun contains(date: LocalDate): Boolean =
            !date.isBefore(start) && !date.isAfter(endInclusive)
    }

    private fun MutableSet<LocalDate>.addRange(start: LocalDate, endInclusive: LocalDate) {
        var current = start
        while (!current.isAfter(endInclusive)) {
            add(current)
            current = current.plusDays(1)
        }
    }
}
